.github/workflows/codeql.yml

```diff
@@ -17,19 +20,65 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       fail-fast: false
+      matrix:
+        language: [ 'cpp' ]
     permissions:
       security-events: write
+      
     steps:
-    - uses: actions/checkout@v3
+    - name: Checkout repository
+      uses: actions/checkout@v3
       with: { fetch-depth: 2 }
 
     - name: Initialize CodeQL
       uses: github/codeql-action/init@v2
       with:
-        languages: cpp
+        languages: ${{ matrix.language }}
+        queries: security-and-quality
+      
     - run: |
         make test CC=gcc ASAN= ASAN_OPTIONS=
         ./test/setup_ga_network.sh && make mip_test CC=gcc ASAN= ASAN_OPTIONS=
 
     - name: Perform CodeQL Analysis
       uses: github/codeql-action/analyze@v2
+      with:
+        category: "/language:${{matrix.language}}"
+        upload: false
+      id: step1
+
+    # Filter out rules with low severity or high false positve rate
+    # Also filter out warnings in third-party code
+    - name: Filter out unwanted errors and warnings
+      uses: advanced-security/filter-sarif@v1
+      with:
+        patterns: |
+          -**:cpp/path-injection
+          -**:cpp/world-writable-file-creation
+          -**:cpp/poorly-documented-function
+          -**:cpp/potentially-dangerous-function
+          -**:cpp/use-of-goto
+          -**:cpp/integer-multiplication-cast-to-long
+          -**:cpp/comparison-with-wider-type
+          -**:cpp/leap-year/*
+          -**:cpp/ambiguously-signed-bit-field
+          -**:cpp/suspicious-pointer-scaling
+          -**:cpp/suspicious-pointer-scaling-void
+          -**:cpp/unsigned-comparison-zero
+          -**/cmake*/Modules/**
+        input: ${{ steps.step1.outputs.sarif-output }}/cpp.sarif
+        output: ${{ steps.step1.outputs.sarif-output }}/cpp.sarif
+
+    - name: Upload CodeQL results to code scanning
+      uses: github/codeql-action/upload-sarif@v2
+      with:
+        sarif_file: ${{ steps.step1.outputs.sarif-output }}
+        category: "/language:${{matrix.language}}"
+
+    - name: Upload CodeQL results as an artifact
+      if: success() || failure()
+      uses: actions/upload-artifact@v3
+      with:
+        name: codeql-results
+        path: ${{ steps.step1.outputs.sarif-output }}
+        retention-days: 5
```

Comment:
![image](https://github.com/cesanta/mongoose/assets/89487381/a8eb753a-f6ed-4a8e-a96d-ae9180cd258d)

This step uploads the results as an artifact under the action and is downloadable for up to 5 days. This acts as an archive and allows for the results to be downloaded.

Reply:
I understand that. Now in coloquial human terms, what is that useful for ? What do I get by doing that ? Will the analyzer perform better ? Will I expose security breachs that should be best handled offline ? There is no need for us to download those, is it ? How will we benefit from that ? Or our customers ? Or any security researchers ?

Reply:
Really no change in analysis. Since the artifact (and Code scanning) can only be accessed by maintainers of the project, it doesn't expose any security breaches. It is up you, the devs, if you want it.